src/main/java/de/tum/cit/aet/artemis/assessment/service/ResultService.java

```diff
@@ -530,31 +542,85 @@ private Result shouldSaveResult(@NotNull Result result, boolean shouldSave) {
+            String[] filterOccurrence) {
+        ProgrammingExercise programmingExercise = programmingExerciseRepository.findWithTestCasesByIdElseThrow(exerciseId);
+
         long distinctResultCount = studentParticipationRepository.countDistinctResultsByExerciseId(exerciseId);
-        Set<ProgrammingExerciseTask> tasks = programmingExerciseTaskService.getTasksWithUnassignedTestCases(exerciseId);
-        List<FeedbackDetailDTO> feedbackDetails = studentParticipationRepository.findAggregatedFeedbackByExerciseId(exerciseId);
-
-        return feedbackDetails.stream().map(detail -> {
-            double relativeCount = (detail.count() * 100.0) / distinctResultCount;
-            int taskNumber = tasks.stream().filter(task -> task.getTestCases().stream().anyMatch(tc -> tc.getTestName().equals(detail.testCaseName()))).findFirst()
-                    .map(task -> tasks.stream().toList().indexOf(task) + 1).orElse(0);
-            return new FeedbackDetailDTO(detail.count(), relativeCount, detail.detailText(), detail.testCaseName(), taskNumber);
+
+        List<String> testCaseNames = programmingExercise.getTestCases().stream().map(ProgrammingExerciseTestCase::getTestName).toList();
+
+        List<ProgrammingExerciseTask> tasks = programmingExerciseTaskService.getTasksWithUnassignedTestCases(exerciseId);
+        Map<String, String> taskNameToIndexMap = new HashMap<>();
+        for (int i = 0; i < tasks.size(); i++) {
+            taskNameToIndexMap.put(tasks.get(i).getTaskName(), String.valueOf(i + 1));
+        }
+        Map<String, String> taskIndexToNameMap = new HashMap<>();
+        for (int i = 0; i < tasks.size(); i++) {
+            taskIndexToNameMap.put(String.valueOf(i + 1), tasks.get(i).getTaskName());
+        }
+
+        List<String> filterTaskNames = filterTasks.stream().map(taskIndexToNameMap::get).filter(Objects::nonNull).toList();
+
+        long minOccurrence = 0;
+        long maxOccurrence = Integer.MAX_VALUE;
+        if (filterOccurrence.length == 2) {
+            minOccurrence = Integer.parseInt(filterOccurrence[0]);
+            maxOccurrence = Integer.parseInt(filterOccurrence[1]);
+        }
+
+        final var pageable = PageUtil.createDefaultPageRequest(search, PageUtil.ColumnMapping.FEEDBACK_ANALYSIS);
+
+        final Page<FeedbackDetailDTO> feedbackDetailPage = studentParticipationRepository.findFilteredFeedbackByExerciseId(exerciseId,
+                StringUtils.isBlank(search.getSearchTerm()) ? "" : search.getSearchTerm().toLowerCase(), filterTestCases, filterTaskNames, minOccurrence, maxOccurrence, pageable);
+
+        List<FeedbackDetailDTO> processedDetails = feedbackDetailPage.getContent().stream().map(detail -> {
+            String taskIndex = taskNameToIndexMap.getOrDefault(detail.taskNumber(), "0");
+            return new FeedbackDetailDTO(detail.count(), (detail.count() * 100.00) / distinctResultCount, detail.detailText(), detail.testCaseName(), taskIndex, "StudentError");
         }).toList();
```

Comment:
ChatGPT propose to use the following code:

// 1. Extract test case names using streams
List<String> testCaseNames = programmingExercise.getTestCases().stream()
        .map(ProgrammingExerciseTestCase::getTestName)  // Extracts the name of each test case
        .toList();  // Converts the stream into a list

// 2. Generate filter task names directly
// We can replace the previous taskIndexToNameMap with direct lookup logic.
// Instead of first mapping index to name and then using the map, we can fetch the task name directly in this stream operation.
List<String> filterTaskNames = filterTasks.stream()
        .map(index -> {
            int idx = Integer.parseInt(index);  // Parse index from the filterTasks list
            return (idx > 0 && idx <= tasks.size()) ? tasks.get(idx - 1).getTaskName() : null;  // Retrieve task name if index is valid
        })
        .filter(Objects::nonNull)  // Remove any null values
        .toList();  // Convert to a list

// 3. Set minOccurrence and maxOccurrence based on filterOccurrence
long minOccurrence = filterOccurrence.length == 2 ? Long.parseLong(filterOccurrence[0]) : 0;  // Default min = 0
long maxOccurrence = filterOccurrence.length == 2 ? Long.parseLong(filterOccurrence[1]) : Integer.MAX_VALUE;  // Default max = MAX_INT

// 4. Create pageable object for pagination
final var pageable = PageUtil.createDefaultPageRequest(search, PageUtil.ColumnMapping.FEEDBACK_ANALYSIS);

// 5. Fetch filtered feedback from the repository
final Page<FeedbackDetailDTO> feedbackDetailPage = studentParticipationRepository.findFilteredFeedbackByExerciseId(
        exerciseId,
        StringUtils.isBlank(search.getSearchTerm()) ? "" : search.getSearchTerm().toLowerCase(),
        filterTestCases,
        filterTaskNames,
        minOccurrence,
        maxOccurrence,
        pageable
);

// 6. Process feedback details directly without using taskNameToIndexMap
// We can perform the task index lookup dynamically instead of using a map.
List<FeedbackDetailDTO> processedDetails = feedbackDetailPage.getContent().stream()
        .map(detail -> {
            // Compute the task index dynamically, no need for a pre-built map.
            String taskIndex = tasks.stream()
                    .filter(task -> task.getTaskName().equals(detail.taskNumber()))
                    .findFirst()
                    .map(task -> String.valueOf(tasks.indexOf(task) + 1))
                    .orElse("0");  // Default to "0" if task not found
            
            // Create and return a new FeedbackDetailDTO object with the necessary fields.
            return new FeedbackDetailDTO(
                    detail.count(),
                    (detail.count() * 100.00) / distinctResultCount,
                    detail.detailText(),
                    detail.testCaseName(),
                    taskIndex,
                    "StudentError"
            );
        })
        .toList();  // Convert the stream to a list
        
        
        Key Optimizations:

	1.	Direct Task Name Lookup: Instead of precomputing the task index-to-name mappings, we directly compute the task names in the stream where they are needed. This avoids storing redundant data in the intermediate maps.
	2.	On-the-Fly Task Index Calculation: In the feedback detail processing, we now compute the task index dynamically by searching through the list of tasks. This eliminates the need for the taskNameToIndexMap.

By removing the maps and using direct lookups, the code becomes more streamlined and memory-efficient, especially if the lists are relatively small. If the size of tasks is large, you may want to consider performance trade-offs, as the stream-based lookup in tasks.stream().filter() will iterate through the list on every access.

Reply:
The lists will be always small and tasks should not have more than 15 in worst case scenarios (at least i have not seen a programming exercise with more than 15 tasks). Therefore using streams should not be an issue. The paged feedback entries are also capped at 20 per page therefore using streams with that is also not an issue.